File: src/recovery_service/stage2/evaluation.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, Field

from .schemas import (
    ActionCandidate,
    CounterfactualSimulation,
    DecisionProposal,
    DiagnosisResult,
    EvidenceManifest,
    FailureDNA,
    IncidentCluster,
    RecoveryEligibility,
    RecoveryGenome,
    ShadowEvaluation,
    TemporalFeatures,
)


class ValueSemantics:
    OBSERVED = "OBSERVED"
    PREDICTED = "PREDICTED"
    PROPOSED = "PROPOSED"
    ESTIMATED = "ESTIMATED"
    VERIFIED = "VERIFIED"
    UNKNOWN = "UNKNOWN"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"
    UNAVAILABLE = "UNAVAILABLE"
    STALE = "STALE"
    BLOCKED = "BLOCKED"


class MetricValue(BaseModel):
    value: Any
    semantic_status: str  # e.g. PREDICTED, VERIFIED, OBSERVED, UNKNOWN
    unit: str = "TEXT"  # CURRENCY_INR, PERCENTAGE, COUNT, TEXT
    source_artifact: str = "UNKNOWN"
    source_version: str = "1.0"
    computed_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class DataQualityStatus(BaseModel):
    recovery_case: bool = False
    evidence_manifest: bool = False
    diagnosis: bool = False
    failure_dna: bool = False
    incident: bool = False
    compliance: bool = False
    genome: bool = False
    proposal: bool = False
    shadow_evaluation: bool = False
    outcome_verified: str = "NOT_YET_OBSERVED"


class CaseEvaluationProjection(BaseModel):
    case_id: str
    merchant_id: str | None
    payment_id: str
    order_id: str | None
    amount: MetricValue
    currency: str
    payment_rail: str
    state: MetricValue
    state_version: int
    recovery_eligible: MetricValue
    
    # Evidence & Timeline
    evidence_manifest: dict[str, Any] | None = None
    
    # Diagnosis & DNA
    diagnosis: dict[str, Any] | None = None
    failure_dna: dict[str, Any] | None = None
    temporal_features: dict[str, Any] | None = None
    
    # Incident & Compliance
    incident: dict[str, Any] | None = None
    compliance: dict[str, Any] | None = None
    
    # RecoveryGenome
    genome: dict[str, Any] | None = None
    
    # Capability & Counterfactuals
    action_capability_matrix: list[dict[str, Any]] = Field(default_factory=list)
    counterfactual_simulations: list[Any] = Field(default_factory=list)
    
    # Decision Proposal & Optimizer
    decision_proposal: dict[str, Any] | None = None
    
    # Shadow Mode Evaluation
    shadow_evaluation: dict[str, Any] | None = None
    
    # GenAI Non-Authoritative Explanation
    genai_explanation: dict[str, Any] | None = None
    
    # Quality & Provenance
    data_quality: DataQualityStatus
    provenance: dict[str, Any] = Field(default_factory=dict)
    evaluated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


def build_metric_value(
    value: Any,
    semantic_status: str,
    unit: str = "TEXT",
    source_artifact: str = "UNKNOWN",
    source_version: str = "1.0",
) -> MetricValue:
    """Construct typed MetricValue preserving strict value semantics."""
    return MetricValue(
        value=value,
        semantic_status=semantic_status,
        unit=unit,
        source_artifact=source_artifact,
        source_version=source_version,
    )

File: frontend/src/types/evaluation.ts
export interface MetricValue {
  value: any;
  value_semantics: 'OBSERVED' | 'PREDICTED' | 'VERIFIED' | 'COUNTERFACTUAL' | 'EVALUATED' | 'GOVERNED';
  source_type: string;
  source_version: string;
}

export interface F4Report {
  report_id?: string;
  merchant_id?: string;
  experiment_id?: string;
  experiment_version?: string;
  status: string; // e.g. "NOT_AVAILABLE", "SUCCESS", "INVALID"
  positivity_status?: string | null;
  reason?: string | null;
  estimand_population?: string | null;
  allocation_proportion_p?: number | null;
  eligible_population_count?: number | null;
  observed_control_count?: number | null;
  observed_treatment_count?: number | null;
  point_estimate_paise_per_unit?: number | null;
  incremental_recovered_revenue_paise?: number | null;
  counterfactual_control_revenue_paise?: number | null;
  standard_error?: number | null;
  confidence_interval_lower?: number | null;
  confidence_interval_upper?: number | null;
  invalidation_reasons?: string[];
  raw_report_json?: Record<string, any> | null;
  evaluated_at?: string | null;
}

export interface CaseEvaluationProjection {
  case_id: string;
  merchant_id: string;
  payment_id: string;
  order_id?: string | null;
  amount: MetricValue;
  currency: string;
  payment_rail: string;
  state: MetricValue;
  state_version: number;
  recovery_eligible: MetricValue;
  evidence_manifest?: Record<string, any> | null;
  diagnosis?: Record<string, any> | null;
  failure_dna?: Record<string, any> | null;
  temporal_features?: Record<string, any> | null;
  incident?: Record<string, any> | null;
  compliance?: Record<string, any> | null;
  genome?: Record<string, any> | null;
  action_capability_matrix?: Array<Record<string, any>>;
  counterfactual_simulations?: Array<Record<string, any>>;
  decision_proposal?: Record<string, any> | null;
  shadow_evaluation?: Record<string, any> | null;
  genai_explanation?: Record<string, any> | null;
  data_quality?: Record<string, boolean>;
  provenance?: Record<string, any>;
}
